[RNMobile] Adjust bottom-sheet

lukewalczak opened the pull request:
## Description

Ref to gb-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1837

That PR is fixing the issue with wrong layout within inserter: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1404

That PR is fixing: https://github.com/wordpress-mobile/WordPress-iOS/issues/13304.

**NOTE:** At the same point, it's disabling swipe-to-close option on Android ( swipe / drag is still available in the top most area - near drag indicator).

## How has this been tested?

1. Run the app on a smaller device or on the regular in horizontal mode
2. User should be able to scroll through all the blocks / add them
3. Please check also other bottom-sheets in the app e.g: `Image` settings

## Screenshots <!-- if applicable -->

ios | android
--- | ---
![ios_scroll_swipe](https://user-images.githubusercontent.com/22746080/73434865-d7c56b80-4347-11ea-8875-8ae37756ecb1.gif) | ![android_drag](https://user-images.githubusercontent.com/22746080/73435214-823d8e80-4348-11ea-96b9-f4dcad19ec3a.gif)



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

lukewalczak committed:
Initial work on iOS Inserter

lukewalczak committed:
Merge branch 'master' into rnmobile/inserter

lukewalczak committed:
Merge branch 'master' into rnmobile/inserter

lukewalczak committed:
Disable swiping-to-close on Android

pinarol commented:
cc @iamthomasbishop Could you check the screen captures in this PR?

pinarol commented:
cc @koke could you also verify this is solving the problems you mentioned in the issue?

dratwas commented:
Hi! I was trying to achieve the perfect solution with a scroll view and swipe but failed because of an issue with `react-native-gesture-handler`.  I used `react-native-reanimated-bottom-sheet` that supports all we need (ScrollView inside, swipe to hide, keyboard aware view) but - It is rendered inside our RN root view that doesn't cover the whole screen (we have a native navbar that is above) and it looks like that:
![Simulator Screen Shot - iPhone X - 2020-01-29 at 13 24 33](https://user-images.githubusercontent.com/16336501/73447986-33035800-4360-11ea-8fca-2817d145e65a.png)


We need to render it inside `react-native` Modal component that creates a separate native view that covers the whole screen. Unfortunately, `react-native-gesture-handler` doesn't work on Android in Modal https://github.com/software-mansion/react-native-gesture-handler/issues/139 . Fortunately, @osdnk already created a fix for that :tada: https://github.com/software-mansion/react-native-gesture-handler/pull/937 I tested it and it seems to work. We need to wait for reviews, merge, and release and we will be able to implement the Bottom Sheet that works as we wanted. In addition, we will change the old `PanResponder` that is controlled by the JS responder system to a more performant one.

pinarol commented:
thank you @dratwas for those great news and all your effort on that! 

To sum up, until the time we have that perfect solution we'll need to solve the issue about [block picker not fitting into small screens anymore](https://github.com/wordpress-mobile/WordPress-iOS/issues/13304). So I believe this PR bings a good short term solution to that.

lukewalczak committed:
Merge branch 'master' into rnmobile/inserter

lukewalczak commented on packages/components/src/mobile/bottom-sheet/index.native.js:
```diff
@@ -127,9 +153,15 @@ class BottomSheet extends Component {
+						style={ { maxHeight } }
+						contentContainerStyle={ [ styles.content, contentStyle ] }
+					>
 						{ this.props.children }
```
One thing which is worrying me ( _and I'm not proud of it :(_ ) is placing a `FlatList` within a `ScrollView`. Generally, it shouldn't have any performance issues since scrolling is disabled in a `FlatList`. There is no issue with rewriting it, but we're going to lose nice features such as `ItemSeparatorComponent` or `numColumns`. Please let me know wdyt.

lukewalczak commented:
One more thing which I think it's worth mentioning is these changes affect (fixes) other bottom-sheets in the application, eg: `Image` settings. This is how it looks like on iPhone 6 with the largest font:

before | after
--- | ---
<img width="515" alt="Screenshot 2020-01-30 at 18 25 07" src="https://user-images.githubusercontent.com/22746080/73474148-43cac280-438e-11ea-9ac5-17d8be86f69d.png"> | <img width="515" alt="Screenshot 2020-01-30 at 18 23 21" src="https://user-images.githubusercontent.com/22746080/73474158-46c5b300-438e-11ea-9d10-82c73faa7ac5.png">

As you can notice in the screenshot before the change, there is not too much free space and keep in mind that e.g `Button` block will have more settings inside.

iamthomasbishop commented:
This is _much_ better than what exists now so I agree w/ @pinarol we should ship these improvements and iterate. With that said, I have a few bits of feedback:

* The sizing/spacing feels a bit off on the top drag handle. IIRC, the drag handle area should be ~16px and the scroll area should be flush up against that "container". [Example](https://cloudup.com/c-PPONBQ4fF)

* It looks like something is off about the bottom safe area around the home indicator. Sometimes the contents of the sheet scroll directly beneath the home indicator (which is what I would expect) and sometimes they do not. Is there any way we can remove the background fill from the safe area that covers the sheet contents? Examples: [Expected, not covering the contents](https://cloudup.com/c8Tc3_M_ToR) vs. [Covering](https://cloudup.com/cL-yv_GPiEE)

> NOTE: At the same point, it's disabling swipe-to-close option on Android ( swipe / drag is still available in the top most area - near drag indicator).

Can the user tap the scrim to dismiss the sheet? I would hope so. Even if they can, because we won't provide swipe-to-dismiss on Android, we might want to target that for a shorter max-height on Android so that the scrim and drag handle are slightlymore reachable while still showing ~2.5 rows of blocks — maybe ~250-260px tall so that it [looks something like this](https://cloudup.com/c-i0UgDDhIC)?

lukewalczak committed:
Add listener for device rotation and small styling changes

lukewalczak committed:
Merge branch 'master' into rnmobile/inserter

lukewalczak commented:
**UPDATES**:

* Horizontal:

top | during scroll | bottom
--- | --- | ---
![Screenshot 2020-01-31 at 11 38 55](https://user-images.githubusercontent.com/22746080/73533405-06624580-441f-11ea-9f11-a9d29d0c530b.png) | ![Screenshot 2020-01-31 at 11 38 59](https://user-images.githubusercontent.com/22746080/73533390-fe0a0a80-441e-11ea-8fcd-9300bdd04dd1.png) | ![Screenshot 2020-01-31 at 11 39 04](https://user-images.githubusercontent.com/22746080/73533376-f6e2fc80-441e-11ea-865d-0a8ab44f3b19.png)

* Vertical:

top | during scroll | bottom
--- | --- | ---
![Screenshot 2020-01-31 at 11 38 35](https://user-images.githubusercontent.com/22746080/73533470-3ad60180-441f-11ea-93c3-a531cc191ab7.png) | ![Screenshot 2020-01-31 at 11 38 46](https://user-images.githubusercontent.com/22746080/73533475-3d385b80-441f-11ea-933f-2c3ad7b51fd4.png) | ![Screenshot 2020-01-31 at 11 38 49](https://user-images.githubusercontent.com/22746080/73533485-3f021f00-441f-11ea-8a15-1cfbf96497a1.png)

lukewalczak committed:
Merge branch 'master' into rnmobile/inserter

lukewalczak committed:
Small cleanup

lukewalczak committed:
More refactor

etoledom commented on packages/components/src/mobile/bottom-sheet/index.native.js:
```diff
@@ -118,17 +155,29 @@ class BottomSheet extends Component {
+						contentContainerStyle={ [
+							styles.content,
+							hideHeader && styles.emptyHeader,
+							{ paddingBottom: this.state.safeAreaBottomInset },
```
This seems to not be doing much. It seems like the space between the last button and the home indicator is quite small. I tested removing/changing this line without any effect.

Again, I trust @iamthomasbishop so check if this extra space is really necessary.

lukewalczak replied:
Fixed, it looks the same as before fix!

etoledom reviewed (commented):
Looks great! I think that the current disadvantages are a good trade-off for solving the main concern we had from a long time (the scrolling capability), and the problems that the lack of it was bringing us in.

I'm really happy to see this! Great job 🎉 !

I added a few questions and concerns as code comments, but nothing really to worry about.
Beside that, I saw these two details:

- On Android I still see a two columns inserter:

<img src="https://user-images.githubusercontent.com/9772967/73554841-c3b76200-444c-11ea-9913-033f91bd1220.png"  width="300">

- There seems to be bouncing even when the content is smaller than the maximum height:

![bouncing](https://user-images.githubusercontent.com/9772967/73554954-fe20ff00-444c-11ea-9022-0ebc83d56f4e.gif)

Apart of that, some of the requests from @iamthomasbishop seems to not be implemented yet. Is that correct? In particular I'm missing the extra space bellow the top handle.

From https://github.com/WordPress/gutenberg/pull/18574#issuecomment-580382194:
>It looks like something is off about the bottom safe area around the home indicator. Sometimes the contents of the sheet scroll directly beneath the home indicator (which is what I would expect) and sometimes they do not. Is there any way we can remove the background fill from the safe area that covers the sheet contents? Examples: Expected, not covering the contents vs. Covering

I wasn't able to reproduce this problem 🤔

iamthomasbishop commented:
@lukewalczak I think on landscape, we can set more than 50% max-height — probably ~90%. For example, some of the system sheets on iOS look like this in landscape:

![IMG_8371](https://user-images.githubusercontent.com/1200432/73567367-e422f480-442b-11ea-9ac4-a0d8a695e2d2.jpeg)

> There seems to be bouncing even when the content is smaller than the maximum height

@etoledom It's hard to tell from the gif above, but does that mean you're not able to swipe to dismiss the sheet?

koke commented:
> cc @koke could you also verify this is solving the problems you mentioned in the issue?

I tested on iPhone 8 with Display Zoom enabled and it's looking pretty good ✨

lukewalczak committed:
Refactor horizontal mode

lukewalczak commented on packages/components/src/mobile/bottom-sheet/index.native.js:
```diff
@@ -62,6 +69,30 @@ class BottomSheet extends Component {
+		const { height } = Dimensions.get( 'window' );
+		const { safeAreaBottomInset } = this.state;
+
+		this.setState( { maxHeight: height / 2 - safeAreaBottomInset } );
```
Fixed, I set 90% of height as Thomas said in the comment below!
<img width="895" alt="Screenshot 2020-02-03 at 13 45 59" src="https://user-images.githubusercontent.com/22746080/73654353-c2c04380-468b-11ea-89ad-a96300b5b32d.png">

lukewalczak commented:
**UPDATES**

* Fixed bottom area inset [screenshot](https://user-images.githubusercontent.com/22746080/73654301-9efcfd80-468b-11ea-8d5e-9a4359255d01.png)

* Fixed landscape mode [screenshot](https://user-images.githubusercontent.com/22746080/73654353-c2c04380-468b-11ea-89ad-a96300b5b32d.png)

**Q&A** 

> On Android I still see a two columns inserter:

This is kind of expected behavior, especially on devices with an 18:9 (or 18,5:9) aspect ratio and enabled screen zoom on max.

Number of columns is calculated based on `Dimensions.get( 'window' ).width / itemWidth`, where `itemWidth` is fixed and value is `120`. The problem here is the `width` from Dimensions which is changing accordingly to screen zoom (the width is decreasing while screen zoom is increasing). That's why e.g on Samsung S8 on neutral zoom we'll see 3 columns (width is ~411), but on max zoom we'll see 2 columns (width is ~320)

neutral zoom | max zoom
--- | ---
![Screenshot_20200203-140451_Gutenberg](https://user-images.githubusercontent.com/22746080/73655485-74f90a80-468e-11ea-8b9d-fca90f65403e.jpg) | ![Screenshot_20200203-140507_Gutenberg](https://user-images.githubusercontent.com/22746080/73655489-762a3780-468e-11ea-8571-cef5a7930920.jpg)

> There seems to be bouncing even when the content is smaller than the maximum height

Tbh, I was not able to reproduce it before refactor and after as well. Please check it one more time, if you will face that issue again, please send me some repro steps along with device configuration / specs 🙏

lukewalczak commented on packages/block-editor/src/components/inserter/menu.native.js:
```diff
@@ -83,38 +86,40 @@ export class InserterMenu extends Component {
-											fill={ modalIconStyle.fill }
-											size={ modalIconStyle.width }
-										/>
+				<TouchableHighlight>
```
It's a bit hard to explain to will use a gif to illustrate it better:

* without `Touchable` or with a `View`: if you will drag an empty space within a bottom-sheet instead of scrolling you can swipe to close the bottom sheet

![without_toucable](https://user-images.githubusercontent.com/22746080/73655906-50e9f900-468f-11ea-9d17-22e977807c42.gif)

* with `Touchable`: dragging an empty space is scrolling instead of swiping

![with_touchable](https://user-images.githubusercontent.com/22746080/73656061-a1f9ed00-468f-11ea-9d00-91e730c1831c.gif)


Let me know if something is unclear with that!

etoledom replied:
Perfect! Thanks for the detailed explanation.

lukewalczak committed:
Merge branch 'master' into rnmobile/inserter

iamthomasbishop commented:
> This is kind of expected behavior, especially on devices with an 18:9 (or 18,5:9) aspect ratio and enabled screen zoom on max.

@lukewalczak I think even with "large" display zoom, we should ensure there are 3 columns per row. If we need to adjust the spacing around grid items, widths, or font-sizes, then let's do that.

iamthomasbishop commented:
Also, the other screenshots above (bottom sheet inset and landscape mode) look great 👍

lukewalczak committed:
Adjust itemWidth for zoomed screen on Android

lukewalczak commented:
**UPDATES** 

Adjusted item width for a zoomed screen on Android:

regular | zoomed
--- | ---
![Screenshot 2020-02-04 at 11 17 24](https://user-images.githubusercontent.com/22746080/73737635-4852fa80-4743-11ea-9d38-15a38d22e52b.png) | ![Screenshot 2020-02-04 at 11 26 58](https://user-images.githubusercontent.com/22746080/73737638-49842780-4743-11ea-9a8c-845785e415fb.png)

etoledom commented:
> **UPDATES**
> 
> Adjusted item width for a zoomed screen on Android:

Looking great on Android now! 🎉 
Thank you for the updates.


> @etoledom It's hard to tell from the gif above, but does that mean you're not able to swipe to dismiss the sheet?

@iamthomasbishop no, it just tries to scroll the content.

> Tbh, I was not able to reproduce it before refactor and after as well. Please check it one more time, if you will face that issue again, please send me some repro steps along with device configuration / specs 🙏

@lukewalczak you are right, it's not the normal behaviour, but a corner case. That's probably why you couldn't reproduce it. I was able to reproduce it on an iPhone XS with these steps:

1. With the device on Landscape, open an image block's settings sheet.
2. Start written a link or an Alt text.
3. Notice that now the sheet is smaller than the available space, so it becomes scrollable (that's good ✅ ). Also notice that the top went out of the screen (not so good, but not alarming).
4. Scroll all the way until the `Clear All Settings` button is visible.
5. With the Clear button visible and the keyboard still present, rotate the device to Portrait mode.
6. Now the Bottom Sheet bounces being smaller than half screen.
7. You can close it and open the same settings, and the problem persist without the need of the previous steps.

This being a corner case, I don't consider it a blocker for this PR, but it would be great to fix it if it's not too complex. 

Let me know!

lukewalczak committed:
Set accessiblity false

lukewalczak committed:
Merge branch 'master' into rnmobile/inserter

lukewalczak committed:
Add disableScrollViewPanResponder

lukewalczak committed:
Adjust BottomSheet height when a keyboard is opened

lukewalczak committed:
Correct removing keyboard listeners

lukewalczak commented:
**UPDATES**

* BottomSheet height fixed along with redundant scrolling

![ezgif com-resize (2)](https://user-images.githubusercontent.com/22746080/73942906-28ad0500-48f0-11ea-8154-5d805c5d48eb.gif)

etoledom reviewed (approved):
The remaining issues seem to be fixed. Looking great to me! 🎉 

Super job @lukewalczak ! I'm so glad we will finally have a scrolling bottom sheet

lukewalczak committed:
Merge branch 'master' into rnmobile/inserter

lukewalczak committed:
Code clean up

lukewalczak committed:
Revert change

lukewalczak committed:
Revert pushed file by mistake

lukewalczak committed:
Code refactor after CR

lukewalczak committed:
Remove dimensions listener on unmount
